### ISSUE TITLE
Regression in `ptr_arg`

### ISSUE THREAD

**Issue by maxwase** [rust-lang/rust-clippy#9676]
### Summary

Clippy had displayed `ptr_arg` warnings before Rust version 1.60

```
➜  expr git:(master) ✗ cargo +1.59 clippy 
    Checking expr v0.1.0 (/home/maksim/dev/my/expr)
warning: writing `&String` instead of `&str` involves a new object where a slice will do
 --> src/lib.rs:1:15
  |
1 | pub fn foo(_: &String) {}
  |               ^^^^^^^ help: change this to: `&str`
  |
  = note: `#[warn(clippy::ptr_arg)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: writing `&Vec<_>` instead of `&[_]` involves one more reference and cannot be used with non-Vec-based slices
 --> src/lib.rs:2:15
  |
2 | pub fn bar(_: &Vec<u8>) {}
  |               ^^^^^^^^ help: change this to: `&[u8]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: `expr` (lib) generated 2 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
➜  expr git:(master) ✗ cargo +1.60 clippy 
    Checking expr v0.1.0 (/home/maksim/dev/my/expr)
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
    ```

### Reproducer

I tried this code:

```rust
fn foo(_: &String) {}
fn bar(_: &Vec<u8>) {}
```

I expected to see this happen:
```
warning: writing `&String` instead of `&str` involves a new object where a slice will do
 --> src/lib.rs:1:15
  |
1 | pub fn foo(_: &String) {}
  |               ^^^^^^^ help: change this to: `&str`
  |
  = note: `#[warn(clippy::ptr_arg)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: writing `&Vec<_>` instead of `&[_]` involves one more reference and cannot be used with non-Vec-based slices
 --> src/lib.rs:2:15
  |
2 | pub fn bar(_: &Vec<u8>) {}
  |               ^^^^^^^^ help: change this to: `&[u8]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
```

Instead, this happened:

No warning

### Version

```text
Working:
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.0

Not working:
rustc 1.60.0 (7737e0b5c 2022-04-04)
binary: rustc
commit-hash: 7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.60.0
LLVM version: 14.0.0
```

**Comment by Jarcho** [comment]
This only happens when the argument has no name. I'm not sure if this should be changed back. The parameter is intentionally unused, which is a good sign that the function signature is important for some reason.

**Comment by maxwase** [comment]
> This only happens when the argument has no name. I'm not sure if this should be changed back. The parameter is intentionally unused, which is a good sign that the function signature is important for some reason.

I can still see the issue even if those parameters are with name and used

**Comment by kraktus** [comment]
I see the warnings as expected, [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d4c3afe9d55b915054075f47f33d1450)


```rust
 fn foo(x: &String) {}
  |           ^^^^^^^ help: change this to: `&str`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
  = note: `#[warn(clippy::ptr_arg)]` on by default

warning: writing `&Vec` instead of `&[_]` involves a new object where a slice will do
 --> src/lib.rs:2:11
  |
2 | fn bar(x: &Vec<u8>) {}
  |           ^^^^^^^^ help: change this to: `&[u8]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
  ```

**Comment by maxwase** [comment]
> I see the warnings as expected, [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d4c3afe9d55b915054075f47f33d1450)
> 
> ```rust
>  fn foo(x: &String) {}
>   |           ^^^^^^^ help: change this to: `&str`
>   |
>   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
>   = note: `#[warn(clippy::ptr_arg)]` on by default
> 
> warning: writing `&Vec` instead of `&[_]` involves a new object where a slice will do
>  --> src/lib.rs:2:11
>   |
> 2 | fn bar(x: &Vec<u8>) {}
>   |           ^^^^^^^^ help: change this to: `&[u8]`
>   |
>   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
> ```

Yeah, in these examples it works, but in [those](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9a798fb3d31e8a199548c140bdb92eab) it is not:

**Comment by maxwase** [comment]
> That's the same as #8995. Generics and trait's are currently assumed to require the current type

Ok, I guess we should generalize that issue more then
